lib/charms/lego_base_k8s/v1/lego_client.py

```diff
@@ -0,0 +1,446 @@
+
+        [
+            self.framework.observe(event, self._configure)
+            for event in [
+                self.on[CA_TRANSFER_RELATION_NAME].relation_joined,
+                self.on[CERTIFICATES_RELATION_NAME].relation_changed,
+                self.on.secret_changed,
+                self.on.config_changed,
+                self.on.update_status,
+            ]
+        ]
+        self.framework.observe(self.on.collect_unit_status, self._on_collect_status)
+
+        self._plugin = plugin
+
+    def _on_collect_status(self, event: CollectStatusEvent) -> None:
+        """Handle the collect status event."""
+        if not self.unit.is_leader():
+            event.add_status(
+                BlockedStatus(
+                    "this charm does not scale, only the leader unit manages certificates."
+                )
+            )
+            return
+        if err := self._validate_charm_config_options():
+            event.add_status(BlockedStatus(err))
+            return
+        if err := self._validate_plugin_config_options(self._plugin_config):
+            event.add_status(BlockedStatus(err))
+            return
+        event.add_status(ActiveStatus(self._get_certificate_fulfillment_status()))
+
+    def _configure(self, event: EventBase) -> None:
+        """Configure the Lego provider."""
+        if not self.unit.is_leader():
+            logger.error("only the leader unit can handle certificate requests")
+            return
+        if err := self._validate_charm_config_options():
+            logger.error(err)
+            return
+        if err := self._validate_plugin_config_options(self._plugin_config):
+            logger.error(err)
+            return
+        self._configure_certificates()
+        self._configure_ca_certificates()
+
+    def _configure_certificates(self):
+        """Attempt to fulfill all certificate requests."""
+        certificate_requests = self.tls_certificates.get_certificate_requests()
+        provided_certificates = self.tls_certificates.get_provider_certificates()
+        certificate_pair_map = {
+            csr: list(
+                filter(
+                    lambda x: x.relation_id == csr.relation_id
+                    and x.certificate_signing_request.raw == csr.certificate_signing_request.raw,
+                    provided_certificates,
+                )
+            )
+            for csr in certificate_requests
+        }
+        with self.maintenance_status("processing certificate requests"):
+            for certificate_request, assigned_certificates in certificate_pair_map.items():
+                if not assigned_certificates:
+                    self._generate_signed_certificate(
+                        csr=certificate_request.certificate_signing_request,
+                        relation_id=certificate_request.relation_id,
+                    )
+
+    def _configure_ca_certificates(self):
+        """Distribute all used CA certificates to requirers."""
```

Comment:
With Let's Encrypt, those will only be the Intermediate CAs, not the root CA, as it is nowhere in the response. Not sure if it's worth noting in the documentation however, as trusting those explicitly should be enough.

Reply:
true, just called it issuer certificate to align with letsencrypt wording.